File: scripts/dft.py
import time

import numpy as np
import scipy

from utils import generate_occupations, occupation_string
from functionals import slater_x, vwn_xc


class SingleAtomDFT:
    def __init__(self, z, r_min, r_max, n_grid, xc_functional=None, occupations=None):
        self.z = z
        # occupation string
        if occupations is None:
            self.occupations = generate_occupations(z)
        else:
            self.occupations = occupation_string(occupations)
        # simpson rule works better with uneven number of samples
        if n_grid % 2 == 0:
            n_grid = n_grid + 1
        self.x = np.linspace(np.log(r_min), np.log(r_max), n_grid)
        self.dx = (self.x[-1] - self.x[0]) / (n_grid - 1)
        self.r = np.exp(self.x)
        self.rho = np.zeros_like(self.r)
        self.xc_functional = xc_functional
        if xc_functional is None:
            self.xc_functional = slater_x
        # used to access energies after completing SCF loop
        self.e_kin = None
        self.e_ext = None
        self.e_xc = None
        self.e_hartree = None
        self.e_total = None
        # histories used for pulay mixing
        self.densities = list()
        self.residuals = list()

    def get_v_external(self):
        e_ext = -4 * np.pi * self.z * scipy.integrate.simpson(self.rho * self.r ** 2, self.x)
        v_ext = -self.z / self.r
        return e_ext, v_ext

    def get_v_hartree(self):
        integral_inner = scipy.integrate.cumulative_simpson(self.rho * self.r ** 3, dx=self.dx, initial=0) / self.r
        integral_outer = scipy.integrate.cumulative_simpson((self.rho * self.r ** 2)[::-1], dx=self.dx, initial=0)[::-1]
        v_hartree = 4 * np.pi * (integral_inner + integral_outer)
        e_hartree = 2 * np.pi * scipy.integrate.simpson(v_hartree * self.rho * self.r ** 3, self.x)
        return e_hartree, v_hartree

    def solve_ks(self, l, v_eff):
        # solve generalized hermitian eigenvalue problem A * chi = epsilon * B * chi
        # with diagonal matrix B
        a_diagonal = 1 / self.dx ** 2 + 1 / 8 + l * (l + 1) / 2 + self.r ** 2 * v_eff
        a_off = np.full(self.r.shape[0] - 1, - 1 / (2 * self.dx ** 2))
        b = self.r ** 2

        # Neumann BC: chi'(x_min) = (l+1/2) * chi(x_min)
        a_diagonal[0] += (l + 0.5) / self.dx
        a_diagonal[0] /= 2
        b[0] /= 2

        # C = B^(-1/2) * A * B^(-1/2)
        c_diagonal = a_diagonal / b
        c_off = a_off / np.sqrt(b[:-1] * b[1:])

        # solve eigenvalue problem C * y = epsilon * y for first n eigenvalues
        # remove last values for Dirichlet boundary conditions
        occupation = len(self.occupations[l])

        epsilon, y = scipy.linalg.eigh_tridiagonal(c_diagonal[:-1], c_off[:-1], select="i",
                                                   select_range=(0, occupation - 1), lapack_driver="stebz")

        # transform back to chi and append zeros for boundary
        chi = np.zeros(shape=(self.r.shape[0], occupation))
        chi[:-1] = y / np.sqrt(b[:-1, np.newaxis])
        # transform back to u
        u = chi * np.sqrt(self.r)[:, np.newaxis]

        # normalize u
        for n in range(occupation):
            norm = scipy.integrate.simpson(u[:, n] ** 2 * self.r, self.x)
            u[:, n] /= np.sqrt(norm)
        return epsilon, u

    def construct_rho(self, eigen_vectors):
        new_rho = np.zeros_like(self.rho)
        for l in range(len(self.occupations)):
            for n in range(len(self.occupations[l])):
                new_rho += self.occupations[l][n] * eigen_vectors[l][:, n] ** 2
        new_rho /= (4 * np.pi * self.r ** 2)
        return new_rho

    def pulay_mixing(self, new_rho, steps=5):
        self.densities.append(new_rho)
        self.residuals.append(new_rho - self.rho)
        if len(self.residuals) >= steps:
            last_residuals = np.array(self.residuals[-steps:])
            weights = self.r ** 3
            weighted_residuals = last_residuals * weights[np.newaxis, :]
            overlap = np.full(shape=(steps + 1, steps + 1), fill_value=-1.0)
            overlap[:steps, :steps] = weighted_residuals @ last_residuals.T
            overlap[-1, -1] = 0
            rhs = np.zeros(steps + 1)
            rhs[-1] = -1
            coefficients = scipy.linalg.solve(overlap, rhs)[:-1]
            last_densities = np.array(self.densities[-steps:])
            return np.tensordot(last_densities, coefficients, (0, 0))
        elif len(self.residuals) > 1:
            return self.rho * 0.8 + new_rho * 0.2
        else:
            return new_rho

    def kinetic_energy(self, eigen_values, v_eff):
        eigen_sum = 0
        for l in range(len(self.occupations)):
            for n in range(len(self.occupations[l])):
                eigen_sum += self.occupations[l][n] * eigen_values[l][n]
        return eigen_sum - 4 * np.pi * scipy.integrate.simpson(v_eff * self.rho * self.r ** 3, self.x)

    def electron_count(self):
        return 4 * np.pi * scipy.integrate.simpson(self.rho * self.r ** 3, self.x)

    def run_scf(self, e_tol=1e-5, rho_tol=1e-3, max_iter=100, min_iter=5):
        v_eff = np.zeros_like(self.r)
        for i in range(max_iter):
            print(f"Iteration {i}")

            # get eigen values and vectors for each l
            eigen_values, eigen_vectors = list(), list()
            for l in range(len(self.occupations)):
                values_l, vectors_l = self.solve_ks(l, v_eff)
                eigen_values.append(values_l)
                eigen_vectors.append(vectors_l)

            # new rho and mixing
            new_rho = self.construct_rho(eigen_vectors)
            rho_diff = np.linalg.norm(new_rho - self.rho)
            self.rho = self.pulay_mixing(new_rho)

            # get energies and v_eff with new rho
            self.e_ext, v_ext = self.get_v_external()
            self.e_hartree, v_hartree = self.get_v_hartree()
            self.e_xc, v_xc = self.xc_functional(self.x, self.r, self.rho)
            v_eff = v_ext + v_xc + v_hartree
            self.e_kin = self.kinetic_energy(eigen_values, v_eff)
            new_e_total = self.e_kin + self.e_ext + self.e_xc + self.e_hartree
            if self.e_total is not None:
                e_total_diff = np.abs(self.e_total - new_e_total)
            else:
                e_total_diff = None
            self.e_total = new_e_total

            print(f"\tElectrons: {self.electron_count(): .5g}\n"
                  f"\tε        : {[a.tolist() for a in eigen_values]}\n"
                  f"\tE_xc     : {self.e_xc: .5g}\n"
                  f"\tE_nucleus: {self.e_ext: .5g}\n"
                  f"\tE_coulomb: {self.e_hartree: .5g}\n"
                  f"\tE_kinetic: {self.e_kin: .5g}\n"
                  f"\tE_total  : {self.e_total: .5g}")
            if e_total_diff is not None:
                print(f"\tDelta E  : {e_total_diff: .5g}\n"
                      f"\tDelta rho: {rho_diff: .5g}")

            # check convergence criteria
            if i >= min_iter and rho_diff < rho_tol and e_total_diff < e_tol:
                print(f"Reached desired convergence. (rho = {rho_diff:.5g}, E = {e_total_diff:.5g})")
                return True, i
        print("Maximum number of iterations reached.")
        return False, max_iter


if __name__ == "__main__":
    start = time.time()
    dft = SingleAtomDFT(1, 1e-4, 100, 3001, xc_functional=vwn_xc)
    dft.run_scf()
    print(f"{time.time() - start:.3f}s elapsed.")

File: src/dft.rs
//! Module containing the main code of the DFT implementation.
//! single_atom_dft() is the main function that should be called to perform
//! the SCF loop. All other functions are called during the SCF loop.

use std::f64::consts;
use crate::integrate;
use crate::linalg;

/// Used to pass the integer occupations to the DFT function.
#[derive(Clone)]
pub struct Occupations {
    // integer occupation number for each (l, n) e.g. [[2, 2], [4], [], []] for Oxygen
    pub f: [Vec<i32>; 4],
    // for how many l is f not empty; e.g. 2 for Oxygen
    pub max_l: usize,
}

/// Contains the grid points of the log grid with n points, where the points in
/// x are equally spaced by h_x and the real space points are r=e^x.
pub struct LogGrid {
    pub n: usize,
    pub x: Vec<f64>,
    pub r: Vec<f64>,
    pub h_x: f64,
}

/// Stores the eigenvalues and eigenvectors for each l.
pub struct Orbitals {
    pub eigenvalues: [Vec<f64>; 4],
    pub eigenvectors: [Vec<f64>; 4],
}

/// All the components of the total energy for a single iteration.
pub struct Energy {
    pub external: f64,
    pub hartree: f64,
    pub xc: f64,
    pub kinetic: f64,
    pub total: f64,
}

/// The return value after the SCF loop converged. Contains the energies,
/// the density, the orbitals etc.
pub struct DFTResult {
    pub energy: Energy,
    pub density: Vec<f64>,
    pub orbitals: Orbitals,
    pub grid: LogGrid,
    pub occupations: Occupations,
    pub iterations: usize,
    pub success: bool,
}

/// Alias for the function pointer used to pass the XC functionals.
pub type XCFunctional = fn(&LogGrid, &Vec<f64>, &mut Vec<f64>) -> f64;

/// Adds the external potential to v_eff and returns the potential energy.
fn add_v_external(z: i32, grid: &LogGrid, rho: &Vec<f64>, v_eff: &mut Vec<f64>) -> f64 {
    // external potential
    for i in 0..grid.n {
        v_eff[i] += -z as f64 / grid.r[i];
    }
    // external energy
    let integrand = (0..grid.n).map(|i| rho[i] * grid.r[i].powi(2)).collect();
    -4.0 * consts::PI * z as f64 * integrate::simpson(&integrand, grid.h_x)
}

/// Adds the hartree potential to v_eff and returns the hartree energy. Uses a
/// Greens function approach to solve the poisson equation.
fn add_v_hartree(grid: &LogGrid, rho: &Vec<f64>, v_eff: &mut Vec<f64>) -> f64 {
    // each point is determined from two integrals
    // using the cumulative integration the integrals are determined for all grid points at once
    let mut inner_integral = (0..grid.n).map(|i| rho[i] * grid.r[i].powi(3)).collect();
    let mut outer_integral = (0..grid.n).map(|i| rho[i] * grid.r[i].powi(2)).rev().collect();
    inner_integral = integrate::cumulative_simpson(&inner_integral, grid.h_x);
    outer_integral = integrate::cumulative_simpson(&outer_integral, grid.h_x);
    // create the hartree potential and the integrand for the hartree energy
    let mut integrand = vec![0.0; grid.n];
    for i in 0..grid.n {
        // the outer integral was reversed to cumulatively integrate it from the
        // back, it needs to be accessed in reverse to fix the order again
        let v_hartree_i = 4.0 * consts::PI *
            (inner_integral[i] / grid.r[i] + outer_integral[grid.n - i - 1]);
        v_eff[i] += v_hartree_i;
        integrand[i] = v_hartree_i * rho[i] * grid.r[i].powi(3);
    }
    2.0 * consts::PI * integrate::simpson(&integrand, grid.h_x)
}

/// Solves the Kohn-Sham equations for the specified l, grid and v_eff.
/// n_shells determines how many eigenvalues and eigenvectors are calculated.
/// They are returned in ascending order: first the eigenvalues, then all
/// eigenvectors concatenated in a single vector.
fn solve_kohn_sham(grid: &LogGrid, l: usize, v_eff: &Vec<f64>, n_shells: usize)
                   -> (Vec<f64>, Vec<f64>) {
    // second-order finite differences lead to a symmetric tridiagonal matrix
    // the equations were adjusted for the log grid r = e^x
    // initialize stencil matrix and right side diagonal matrix B
    // omit last row for Dirichlet boundary at r_max
    let diag = (0..grid.n - 1).map(|i|
        1.0 / grid.h_x.powi(2) + 0.125 + 0.5 * (l * (l + 1)) as f64 + grid.r[i].powi(2) * v_eff[i]
    ).collect();
    let off = vec![-0.5 / grid.h_x.powi(2); grid.n - 2];
    let mut a = linalg::TridiagonalMatrix { n: grid.n - 1, diag, off };
    let mut b: Vec<f64> = (0..grid.n - 1).map(|i| grid.r[i].powi(2)).collect();
    // Neumann BC at r_min
    a.diag[0] += l as f64 + 0.5 / grid.h_x;
    a.diag[0] /= 2.0;
    b[0] /= 2.0;
    // the LAPACK function does not work with a generalized problem A * x = epsilon * B * x
    // instead solve C * y = epsilon * y, transform C = B^(-1/2) * A * B^(-1/2) in-place
    for i in 0..grid.n - 2 {
        a.diag[i] /= b[i];
        a.off[i] /= (b[i] * b[i + 1]).sqrt();
    }
    a.diag[grid.n - 2] /= b[grid.n - 2];

    // solve eigenvalue problem
    let (mut epsilon, y) = linalg::eigh_tridiagonal(a, n_shells as i32);

    // sort eigenvalues
    let mut sort_index = (0..n_shells).collect::<Vec<_>>();
    sort_index.sort_by(|&a, &b| epsilon[a].partial_cmp(&epsilon[b]).unwrap());
    epsilon = sort_index.iter().map(|&i| epsilon[i]).collect();

    // sort eigenvectors and append 0.0 boundary point
    let mut u = vec![0.0; grid.n * n_shells];
    for dst in 0..n_shells {
        let src = sort_index[dst];
        for i in 0..grid.n - 1 {
            // destination vectors are 1 longer than source vectors
            let src_index = src * (grid.n - 1) + i;
            let dst_index = dst * grid.n + i;
            // transform back from y to u during sort
            u[dst_index] = y[src_index] / b[i].sqrt() * grid.r[i].sqrt();
        }
    }

    // normalize u
    for j in 0..n_shells {
        let integrand = (0..grid.n).map(|i| u[j * grid.n + i].powi(2) * grid.r[i]).collect();
        let norm = integrate::simpson(&integrand, grid.h_x).sqrt();
        for i in 0..grid.n {
            u[j * grid.n + i] /= norm;
        }
    }
    (epsilon, u)
}

/// Constructs and returns the density from the given eigenvectors.
fn construct_rho(grid: &LogGrid, u: &[Vec<f64>; 4], occupations: &Occupations) -> Vec<f64> {
    let mut new_rho = vec![0.0; grid.n];
    for l in 0..occupations.max_l {
        for n in 0..occupations.f[l].len() {
            for i in 0..grid.n {
                new_rho[i] += occupations.f[l][n] as f64 * u[l][n * grid.n + i].powi(2) /
                    (4.0 * consts::PI * grid.r[i].powi(2));
            }
        }
    }
    new_rho
}

/// Performs pulay mixing by consuming the old and new densities and returning
/// the resulting density. The histories of the residuals and densities need to
/// be passed and will be modified to contain the new residual and density.
fn pulay_mixing(grid: &LogGrid, densities: &mut Vec<Vec<f64>>, residuals: &mut Vec<Vec<f64>>,
                old_rho: Vec<f64>, mut new_rho: Vec<f64>, steps: usize, i: usize) -> Vec<f64> {
    // calculate residual and save residual and density in history
    let residual = (0..grid.n).map(|i| new_rho[i] - old_rho[i]).collect();
    residuals[i % steps] = residual;
    densities[i % steps] = new_rho.clone();
    if i == 0 {
        // in first iteration just leave density as is
    } else if i < steps - 1 {
        // use linear mixing until history is filled
        for j in 0..grid.n {
            new_rho[j] = old_rho[j] * 0.8 + new_rho[j] * 0.2;
        }
    } else {
        // pulay mixing using history
        // create linear system Ax=b
        let mut a = vec![-1.0; (steps + 1).pow(2)];
        let mut b = vec![0.0; steps + 1];
        a[(steps + 1).pow(2) - 1] = 0.0;
        b[steps] = -1.0;
        for x in 0..steps {
            // only fill lower triangle of A
            for y in x..steps {
                // vector dot product of residuals, weighted by r^3
                let a_i = x * (steps + 1) + y;
                a[a_i] = (0..grid.n).fold(0.0, |acc, j|
                    acc + grid.r[j].powi(3) * residuals[x][j] * residuals[y][j]
                );
            }
        }
        // solve linear system
        let ls = linalg::LinearSystem { n: steps + 1, a, b };
        let coefficients = linalg::solve_symmetric(ls, false);
        // linear superposition of densities using coefficients
        new_rho.fill(0.0);
        for j in 0..grid.n {
            for k in 0..steps {
                new_rho[j] += coefficients[k] * densities[k][j];
            }
        }
    }
    new_rho
}

/// Returns the kinetic energy for the given density, v_eff and eigenvalues.
fn get_kinetic_energy(grid: &LogGrid, rho: &Vec<f64>, v_eff: &Vec<f64>, eigenvalues: &[Vec<f64>; 4],
                      occupations: &Occupations) -> f64 {
    let mut eigenvalue_sum = 0.0;
    for l in 0..occupations.max_l {
        for n in 0..occupations.f[l].len() {
            eigenvalue_sum += occupations.f[l][n] as f64 * eigenvalues[l][n];
        }
    }
    let integrand = (0..grid.n).map(|i| v_eff[i] * rho[i] * grid.r[i].powi(3) ).collect();
    eigenvalue_sum - 4.0 * consts::PI * integrate::simpson(&integrand, grid.h_x)
}

/// Returns the electron count by integrating over the given density.
fn get_electron_count(grid: &LogGrid, rho: &Vec<f64>) -> f64 {
    let integrand = (0..grid.n).map(|i| rho[i] * grid.r[i].powi(3)).collect();
    4.0 * consts::PI * integrate::simpson(&integrand, grid.h_x)
}

/// Performs DFT for the specified parameters by starting an SCF loop until the
/// convergence criteria are met and returns a struct containing the final
/// density, energies and orbitals.
pub fn single_atom_dft(z: i32, r_min: f64, r_max: f64, n_grid: usize, occupations: Occupations,
                       mixing_steps: usize, max_iter: usize, min_iter: usize, e_tol: f64,
                       rho_tol: f64, xc_functional: XCFunctional, verbose: i32) -> DFTResult {
    // initialize grid
    // uniform grid in x from ln(r_min) to ln(r_max) with spacing h_x
    // log grid in r with r = e^x
    let x: Vec<f64> = (0..n_grid).map(|i|
        r_min.ln() + i as f64 / (n_grid - 1) as f64 * (r_max.ln() - r_min.ln())
    ).collect();
    let r = x.iter().map(|&x| consts::E.powf(x)).collect();
    let h_x = (x[x.len() - 1] - x[0]) / (n_grid - 1) as f64;
    let grid = LogGrid { n: n_grid, x, r, h_x };

    // density and residual history
    let mut density_history = vec![vec![0.0; grid.n]; mixing_steps];
    let mut residual_history = vec![vec![0.0; grid.n]; mixing_steps];
    // initial density and potential
    let mut rho = vec![0.0; n_grid];
    let mut v_eff = vec![0.0; n_grid];

    // to save orbitals and energies each iteration
    let mut orbitals = Orbitals {
        eigenvalues: [vec![], vec![], vec![], vec![]],
        eigenvectors: [vec![], vec![], vec![], vec![]],
    };
    let mut energy = Energy { external: 0.0, hartree: 0.0, xc: 0.0, kinetic: 0.0, total: 0.0 };

    // main SCF loop
    for i in 0..max_iter {
        // solve Kohn-Sham equations and construct new density
        for l in 0..occupations.max_l {
            (orbitals.eigenvalues[l], orbitals.eigenvectors[l]) =
                solve_kohn_sham(&grid, l, &v_eff, occupations.f[l].len());
        }
        let new_rho = construct_rho(&grid, &orbitals.eigenvectors, &occupations);

        // calculate difference to previous density, then apply mixing
        let rho_diff = (0..grid.n).fold(0.0, |diff, j|
            diff + (new_rho[j] - rho[j]).powi(2)
        ).sqrt();
        rho = pulay_mixing(&grid, &mut density_history, &mut residual_history, rho, new_rho,
                           mixing_steps, i);

        // get energies and new effective potential
        v_eff.fill(0.0);
        energy.external = add_v_external(z, &grid, &rho, &mut v_eff);
        energy.hartree = add_v_hartree(&grid, &rho, &mut v_eff);
        energy.xc = xc_functional(&grid, &rho, &mut v_eff);
        energy.kinetic = get_kinetic_energy(&grid, &rho, &v_eff, &orbitals.eigenvalues,
                                            &occupations);

        // calculate energy difference to previous iteration
        let new_e_total = energy.external + energy.kinetic + energy.hartree + energy.xc;
        let e_total_diff = (new_e_total - energy.total).abs();
        energy.total = new_e_total;

        // print
        if verbose >= 2 {
            let electron_count = get_electron_count(&grid, &rho);
            println!("Iteration {i}:\n\
                     \tElectrons: {:12.5}\n\
                     \tε        : {:?}\n\
                     \tE_xc     : {:12.5}\n\
                     \tE_nucleus: {:12.5}\n\
                     \tE_coulomb: {:12.5}\n\
                     \tE_kinetic: {:12.5}\n\
                     \tE_total  : {:12.5}",
                     electron_count, orbitals.eigenvalues, energy.xc, energy.external,
                     energy.hartree, energy.kinetic, energy.total);
        }
        // check convergence criteria
        if i > min_iter && e_total_diff < e_tol && rho_diff < rho_tol {
            if verbose >= 2 {
                println!("Reached desired convergence. (rho = {:.5e}, E = {:.5e})",
                         rho_diff, e_total_diff);
            } else if verbose == 1 {
                println!("Z = {:3}: {:4} {:12.5} {:12.5} {:12.5} {:12.5} {:12.5}    {:?}",
                         z, i, energy.total, energy.kinetic, energy.hartree, energy.external,
                         energy.xc, orbitals.eigenvalues);
            }
            return DFTResult { energy, density: rho, orbitals, grid, occupations, iterations: i,
                success: true };
        }
    }
    if verbose >= 2 { println!("Maximum number of iterations reached.") };
    DFTResult { energy, density: rho, orbitals, grid, occupations, iterations: max_iter,
        success: false }
}
